rfcs/2023-12-25-proxy-implementation.md

```diff
@@ -0,0 +1,108 @@
+
+Each proxy-to-relay (P2R) connection aggregates traffic from multiple clients and protected by a
+
+A proxy runs a barebones SMP protocol towards each active relay, fronted by an `MVar` which, unlike other concurrency primitives, guarantee fairness. With it, each client "forwarding" worker takes a unit of work and tries to submit it to a relay worker.
+
+A client should not be able to inspect/probe relay connections and distinguish a fresh connection from an existing one. Thus each client gets a `server_id` map with ids distinct from the proxy internal relay map. This also precludes accidental coupling between implementation and protocol format with its (quite severe) limits.
+
+### Relay server
+
+As with notification subscriber "clients" a proxy is just an another client with a special set of commands for setup and forwarding.
+The usual SMP precautions apply.
+
+## Failure modes
+
+### Delivery problems
+
+A delivery latency may increase due to:
+- Slow client/proxy handshake.
+- Slow proxy/relay handshake for new servers.
+- Long time waiting for its turn to send due to lots of active clients.
+- Long time to deliver to relay due to overloaded server.
+This can be mitigated by proxy pool.
+A pool may be monitoring proxy latency in general and towards a specific server, then route accordingly.
+
+A message non-delivery may happen due to:
+- Proxy misconfiguration. Report error to user.
+- Relay stopped operating or proxy lost its connection. Signalled to client by proxy `ERR`. Retry later.
+- Proxy stopped operating or client lost its connection. Signalled by connection error. Retry with another proxy or later.
+
+A duplicate delivery may happen due to:
+- Client used multiple routes to deliver (e.g. because of apparent non-delivery).
+- Proxy replaying messages (maliciously or by misconfiguration).
+This is mitigated client side by deduplication. And, optionally, relay side, by using bloom filter with message DH keys.
+
+### Connectivity problems
+
+A proxy may lose connection to a relay.
+The protocol lacks proxy notifications, so a client may only be informed of the failure state when it tries to forward a message.
+If a client tries to send to a dropped connection it gets an `ERR` and must de-associate the server from the proxy.
+The proxy should flush its client queues related to the server and respond with `ERR`.
+The proxy should attempt to reconnect if it has pending messages, as the clients have already got their `OK`.
+Otherwise the proxy may attempt to reconnect to the relay, expecting future client attempts to restore their connections.
+
+A client may lose connection to a proxy.
+It should wipe its `server_id` mapping and reconnect to resume operations.
+A pool may down-weight the proxy proportionally to its uptime to prevent latency problems.
+
+### Denial of service attacks
+
+A malicious client can request connections (esp. to a coopted relay) to exhaust proxy resources.
```

Comment:
coopted means what here? assuming collusion between client and relay, how does this collusion help the attack?